File: core/src/main/java/ir/logicfan/core/di/module/BaseModule.java
package ir.logicfan.core.di.module;

import android.app.Application;
import android.content.Context;
import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import io.reactivex.disposables.CompositeDisposable;
import ir.logicfan.core.data.pref.BasePrefManager;
import ir.logicfan.core.di.qulifier.ApplicationContext;

@Module
public abstract class BaseModule {

    @Binds
    @ApplicationContext
    abstract Context context(Application application);

    @Provides
    static CompositeDisposable disposable() {
        return new CompositeDisposable();
    }

    @Provides
    static BasePrefManager basePrefManager(@ApplicationContext Context context) {
        return new BasePrefManager(context);
    }
}

File: core/src/main/java/ir/logicfan/core/di/module/BaseModule.kt
/*
 * Copyright (C) Logicfan, Inc - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited Proprietary and confidential
 * Written by Mahdi Javaheri Saber <m.javaherisaber@gmail.com>, Mohammad Hosein Abedini <mohammadhoseinabedini@gmail.com>  2020.
 */

package ir.logicfan.core.di.module

import android.app.Application
import android.content.Context
import androidx.lifecycle.ViewModelProvider
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.Reusable
import io.reactivex.disposables.CompositeDisposable
import ir.logicfan.core.di.base.ViewModelFactory

@Module
abstract class BaseModule {

    @Binds
    @Reusable
    abstract fun context(application: Application): Context

    @Binds
    abstract fun viewModelFactory(viewModelFactory: ViewModelFactory): ViewModelProvider.Factory

    @Module
    companion object {
        @Provides
        @JvmStatic
        fun compositeDisposable(): CompositeDisposable = CompositeDisposable()
    }
}
